File: scripts/resolve.py
#!/usr/bin/env python3
"""Resolve roles.json into skill lists and skillOverrides maps.

Single dependency-free helper shared by install.sh (and usable in CI). All
commands read roles.json from the repo root (the directory above scripts/),
or from $ROLES_JSON if set.

Usage:
  resolve.py roles                       List role keys + labels (TSV).
  resolve.py skills <role>               Resolved skill set for a role (one per line).
  resolve.py label  <role>               Human label for a role.
  resolve.py overrides [role]            Emit a skillOverrides JSON object marking
                                         every installed skill outside (pinned ∪ role)
                                         as "name-only". Installed skill names are read
                                         from stdin (newline-separated). No role => only
                                         the pinned set stays on (the baseline).
  resolve.py apply <settings.local.json> <skills_dir> [role]
                                         Merge the name-only baseline (for [role], or
                                         pinned-only) into the settings file's
                                         skillOverrides, preserving all other keys and
                                         any overrides for non-installed skills.
  resolve.py validate <skills_dir>       Integrity check: every referenced skill
                                         exists under <skills_dir>, and every
                                         non-meta skill belongs to >=1 role.
"""
import json
import os
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
ROLES_JSON = os.environ.get("ROLES_JSON", os.path.join(HERE, "..", "roles.json"))


def load():
    with open(ROLES_JSON, encoding="utf-8") as fh:
        return json.load(fh)


def die(msg, code=1):
    print(f"resolve.py: {msg}", file=sys.stderr)
    sys.exit(code)


def role_or_die(data, role):
    roles = data.get("roles", {})
    if role not in roles:
        die(f"unknown role '{role}' (known: {', '.join(sorted(roles))})")
    return roles[role]


def resolved_skills(data, role):
    """Role working set = its core set UNION its own skills, order-stable."""
    r = role_or_die(data, role)
    core = data.get("core", {}).get(r.get("core", "universal"), [])
    seen, out = set(), []
    for skill in list(core) + list(r.get("skills", [])):
        if skill not in seen:
            seen.add(skill)
            out.append(skill)
    return out


def cmd_roles(data, args):
    for key, r in data.get("roles", {}).items():
        print(f"{key}\t{r.get('label', key)}")


def cmd_skills(data, args):
    if not args:
        die("skills requires a <role>")
    for skill in resolved_skills(data, args[0]):
        print(skill)


def cmd_label(data, args):
    if not args:
        die("label requires a <role>")
    print(role_or_die(data, args[0]).get("label", args[0]))


def keep_on_set(data, role):
    """Skills that stay `on` (full description, auto-trigger): the pinned set,
    plus the active role's working set when a role is given."""
    keep = set(data.get("pinned", []))
    if role:
        keep |= set(resolved_skills(data, role))
    return keep


def nameonly_map(data, installed, role):
    keep_on = keep_on_set(data, role)
    return {s: "name-only" for s in installed if s not in keep_on}


def cmd_overrides(data, args):
    role = args[0] if args else None
    if role:
        role_or_die(data, role)
    installed = [ln.strip() for ln in sys.stdin if ln.strip()]
    print(json.dumps(nameonly_map(data, installed, role), indent=2, sort_keys=True))


def cmd_apply(data, args):
    if len(args) < 2:
        die("apply requires <settings.local.json> <skills_dir> [role]")
    settings_path, skills_dir = args[0], args[1]
    role = args[2] if len(args) > 2 else None
    if role and role not in ("all", "none"):
        role_or_die(data, role)
    if role in ("all", "none"):
        role = None

    installed = sorted(
        e for e in os.listdir(skills_dir)
        if os.path.isdir(os.path.join(skills_dir, e))
    )
    desired = nameonly_map(data, installed, role)  # {skill: "name-only"} for the tail

    # Read existing settings (tolerate missing / empty / malformed-as-empty).
    settings = {}
    if os.path.isfile(settings_path):
        try:
            with open(settings_path, encoding="utf-8") as fh:
                settings = json.load(fh) or {}
        except (OSError, ValueError):
            settings = {}

    existing = settings.get("skillOverrides", {})
    if not isinstance(existing, dict):
        existing = {}
    # Preserve overrides for skills we don't manage (not installed here); we own
    # every installed skill's entry: name-only for the tail, absent (=> on) otherwise.
    merged = {k: v for k, v in existing.items() if k not in installed}
    merged.update(desired)
    settings["skillOverrides"] = merged

    os.makedirs(os.path.dirname(os.path.abspath(settings_path)), exist_ok=True)
    with open(settings_path, "w", encoding="utf-8") as fh:
        json.dump(settings, fh, indent=2, sort_keys=True)
        fh.write("\n")
    on_count = len(installed) - len(desired)
    print(f"applied: {len(desired)} name-only, {on_count} on "
          f"(role={role or 'baseline'}) -> {settings_path}")


def cmd_validate(data, args):
    if not args:
        die("validate requires a <skills_dir>")
    skills_dir = args[0]
    errors = []

    # Collect every skill referenced anywhere in the SSOT.
    referenced = set(data.get("pinned", [])) | set(data.get("meta_only", []))
    for core in data.get("core", {}).values():
        referenced |= set(core)
    in_a_role = set()
    for role in data.get("roles", {}):
        skills = set(resolved_skills(data, role))
        referenced |= skills
        in_a_role |= skills

    # Every referenced skill must exist on disk.
    for skill in sorted(referenced):
        if not os.path.isdir(os.path.join(skills_dir, skill)):
            errors.append(f"referenced skill missing on disk: {skill}")

    # Every non-meta skill on disk must belong to >=1 role (no orphans).
    meta = set(data.get("meta_only", []))
    universal = set(data.get("core", {}).get("universal", []))
    technical = set(data.get("core", {}).get("technical", []))
    covered = in_a_role | universal | technical
    for entry in sorted(os.listdir(skills_dir)):
        if not os.path.isdir(os.path.join(skills_dir, entry)):
            continue
        if entry in meta:
            continue
        if entry not in covered:
            errors.append(f"orphan skill (in no role/core): {entry}")

    if errors:
        for e in errors:
            print(f"FAIL: {e}", file=sys.stderr)
        sys.exit(1)
    print(f"OK: roles.json integrity verified against {skills_dir}")


COMMANDS = {
    "roles": cmd_roles,
    "skills": cmd_skills,
    "label": cmd_label,
    "overrides": cmd_overrides,
    "apply": cmd_apply,
    "validate": cmd_validate,
}


def main():
    if len(sys.argv) < 2 or sys.argv[1] not in COMMANDS:
        die(f"usage: resolve.py <{'|'.join(COMMANDS)}> [args]", code=2)
    COMMANDS[sys.argv[1]](load(), sys.argv[2:])


if __name__ == "__main__":
    main()

File: scripts/resolve.mjs
#!/usr/bin/env node
// Resolve roles.json into skill lists and skillOverrides maps.
//
// Dependency-free helper shared by install.mjs / uninstall.mjs (import) and by the
// SessionStart hook + the /role command (CLI). Pure JSON + set logic — runs on
// Linux, macOS, and Windows with only the Node that Claude Code already requires.
//
// Usage (CLI):
//   node resolve.mjs roles                       List role keys + labels (TSV).
//   node resolve.mjs skills <role>               Resolved skill set for a role (one per line).
//   node resolve.mjs label  <role>               Human label for a role.
//   node resolve.mjs overrides [role]            Emit a skillOverrides JSON object marking
//                                                every installed skill outside (pinned ∪ role)
//                                                as "name-only". Installed skill names are read
//                                                from stdin (newline-separated). No role => only
//                                                the pinned set stays on (the baseline).
//   node resolve.mjs apply <settings.local.json> <skills_dir> [role]
//                                                Merge the name-only baseline (for [role], or
//                                                pinned-only) into the settings file's
//                                                skillOverrides, preserving all other keys and
//                                                any overrides for non-installed skills.
//   node resolve.mjs prune <settings.local.json> [skill ...]
//                                                Remove the given skills from the settings
//                                                file's skillOverrides (used by uninstall).
//                                                Drops skillOverrides entirely when empty.
//   node resolve.mjs validate <skills_dir>       Integrity check: every referenced skill
//                                                exists, and every non-meta skill is in >=1 role.
//
// roles.json is read from $ROLES_JSON if set, else ../roles.json relative to this file.

import { readFileSync, writeFileSync, mkdirSync, existsSync, statSync, readdirSync } from "node:fs";
import { dirname, join, resolve as resolvePath } from "node:path";
import { fileURLToPath } from "node:url";

const HERE = dirname(fileURLToPath(import.meta.url));
export const DEFAULT_ROLES_JSON =
  process.env.ROLES_JSON || join(HERE, "..", "roles.json");

// ---- core resolution -------------------------------------------------------

export function loadRoles(rolesPath = DEFAULT_ROLES_JSON) {
  return JSON.parse(readFileSync(rolesPath, "utf-8"));
}

function roleOrDie(data, role) {
  const roles = data.roles || {};
  if (!(role in roles)) {
    const known = Object.keys(roles).sort().join(", ");
    die(`unknown role '${role}' (known: ${known})`);
  }
  return roles[role];
}

// Role working set = its core set UNION its own skills, order-stable.
export function resolvedSkills(data, role) {
  const r = roleOrDie(data, role);
  const core = (data.core || {})[r.core || "universal"] || [];
  const seen = new Set();
  const out = [];
  for (const skill of [...core, ...(r.skills || [])]) {
    if (!seen.has(skill)) {
      seen.add(skill);
      out.push(skill);
    }
  }
  return out;
}

// Skills that stay `on` (full description, auto-trigger): the pinned set, plus the
// active role's working set when a role is given.
export function keepOnSet(data, role) {
  const keep = new Set(data.pinned || []);
  if (role) for (const s of resolvedSkills(data, role)) keep.add(s);
  return keep;
}

export function nameonlyMap(data, installed, role) {
  const keepOn = keepOnSet(data, role);
  const map = {};
  for (const s of installed) if (!keepOn.has(s)) map[s] = "name-only";
  return map;
}

// Directory entries that are skills (subdirectories), sorted.
export function installedSkills(skillsDir) {
  return readdirSync(skillsDir)
    .filter((e) => statSync(join(skillsDir, e)).isDirectory())
    .sort();
}

// ---- settings I/O ----------------------------------------------------------

// Recursively sort object keys so output is deterministic (mirrors Python's
// json.dump(..., sort_keys=True)).
function sortKeysDeep(value) {
  if (Array.isArray(value)) return value.map(sortKeysDeep);
  if (value && typeof value === "object") {
    const out = {};
    for (const k of Object.keys(value).sort()) out[k] = sortKeysDeep(value[k]);
    return out;
  }
  return value;
}

export function loadSettings(settingsPath) {
  if (existsSync(settingsPath) && statSync(settingsPath).isFile()) {
    try {
      return JSON.parse(readFileSync(settingsPath, "utf-8")) || {};
    } catch {
      return {};
    }
  }
  return {};
}

export function writeSettings(settingsPath, settings) {
  mkdirSync(dirname(resolvePath(settingsPath)), { recursive: true });
  writeFileSync(settingsPath, JSON.stringify(sortKeysDeep(settings), null, 2) + "\n");
}

// ---- apply / prune ---------------------------------------------------------

// Merge the name-only baseline into a settings file. Returns {nameOnly, on, role}.
export function applyBaseline(data, settingsPath, skillsDir, role) {
  if (role === "all" || role === "none") role = null;
  if (role) roleOrDie(data, role);

  const installed = installedSkills(skillsDir);
  const desired = nameonlyMap(data, installed, role); // {skill: "name-only"} tail

  const settings = loadSettings(settingsPath);
  let existing = settings.skillOverrides;
  if (!existing || typeof existing !== "object" || Array.isArray(existing)) existing = {};

  // Preserve overrides for skills we don't manage (not installed here); we own every
  // installed skill's entry: name-only for the tail, absent (=> on) otherwise.
  const installedSet = new Set(installed);
  const merged = {};
  for (const [k, v] of Object.entries(existing)) if (!installedSet.has(k)) merged[k] = v;
  Object.assign(merged, desired);
  settings.skillOverrides = merged;

  writeSettings(settingsPath, settings);
  const nameOnly = Object.keys(desired).length;
  return { nameOnly, on: installed.length - nameOnly, role };
}

// Remove the given skills from a settings file's skillOverrides. Returns removed count.
export function pruneSettings(settingsPath, skills) {
  if (!(existsSync(settingsPath) && statSync(settingsPath).isFile())) return { removed: 0, missing: true };
  const settings = loadSettings(settingsPath);
  const overrides = settings.skillOverrides;
  if (!overrides || typeof overrides !== "object" || Array.isArray(overrides)) {
    return { removed: 0, noOverrides: true };
  }
  const want = new Set(skills);
  let removed = 0;
  for (const s of want) {
    if (s in overrides) {
      delete overrides[s];
      removed++;
    }
  }
  if (Object.keys(overrides).length > 0) settings.skillOverrides = overrides;
  else delete settings.skillOverrides; // drop the empty key entirely
  writeSettings(settingsPath, settings);
  return { removed };
}

// ---- validate --------------------------------------------------------------

export function validate(data, skillsDir) {
  const errors = [];

  // Every skill referenced anywhere in the SSOT.
  const referenced = new Set([...(data.pinned || []), ...(data.meta_only || [])]);
  for (const core of Object.values(data.core || {})) for (const s of core) referenced.add(s);
  const inARole = new Set();
  for (const role of Object.keys(data.roles || {})) {
    for (const s of resolvedSkills(data, role)) {
      referenced.add(s);
      inARole.add(s);
    }
  }

  // Every referenced skill must exist on disk.
  for (const skill of [...referenced].sort()) {
    if (!isDir(join(skillsDir, skill))) errors.push(`referenced skill missing on disk: ${skill}`);
  }

  // Every non-meta skill on disk must belong to >=1 role (no orphans).
  const meta = new Set(data.meta_only || []);
  const universal = new Set((data.core || {}).universal || []);
  const technical = new Set((data.core || {}).technical || []);
  const covered = new Set([...inARole, ...universal, ...technical]);
  for (const entry of readdirSync(skillsDir).sort()) {
    if (!isDir(join(skillsDir, entry))) continue;
    if (meta.has(entry)) continue;
    if (!covered.has(entry)) errors.push(`orphan skill (in no role/core): ${entry}`);
  }
  return errors;
}

function isDir(p) {
  return existsSync(p) && statSync(p).isDirectory();
}

// ---- CLI -------------------------------------------------------------------

export function die(msg, code = 1) {
  process.stderr.write(`resolve.mjs: ${msg}\n`);
  process.exit(code);
}

function readStdin() {
  try {
    return readFileSync(0, "utf-8");
  } catch {
    return "";
  }
}

const CLI = {
  roles(data) {
    for (const [key, r] of Object.entries(data.roles || {})) {
      process.stdout.write(`${key}\t${r.label || key}\n`);
    }
  },
  skills(data, args) {
    if (!args[0]) die("skills requires a <role>");
    for (const s of resolvedSkills(data, args[0])) process.stdout.write(s + "\n");
  },
  label(data, args) {
    if (!args[0]) die("label requires a <role>");
    process.stdout.write(roleOrDie(data, args[0]).label || args[0]);
    process.stdout.write("\n");
  },
  overrides(data, args) {
    const role = args[0] || null;
    if (role) roleOrDie(data, role);
    const installed = readStdin()
      .split("\n")
      .map((l) => l.trim())
      .filter(Boolean);
    process.stdout.write(JSON.stringify(sortKeysDeep(nameonlyMap(data, installed, role)), null, 2) + "\n");
  },
  apply(data, args) {
    if (args.length < 2) die("apply requires <settings.local.json> <skills_dir> [role]");
    const { nameOnly, on, role } = applyBaseline(data, args[0], args[1], args[2]);
    process.stdout.write(
      `applied: ${nameOnly} name-only, ${on} on (role=${role || "baseline"}) -> ${args[0]}\n`,
    );
  },
  prune(_data, args) {
    if (!args[0]) die("prune requires <settings.local.json> [skill ...]");
    const settingsPath = args[0];
    const res = pruneSettings(settingsPath, args.slice(1));
    if (res.missing) process.stdout.write(`pruned: 0 (no settings file at ${settingsPath})\n`);
    else if (res.noOverrides) process.stdout.write(`pruned: 0 (no skillOverrides) -> ${settingsPath}\n`);
    else process.stdout.write(`pruned: ${res.removed} skillOverrides entries -> ${settingsPath}\n`);
  },
  validate(data, args) {
    if (!args[0]) die("validate requires a <skills_dir>");
    const errors = validate(data, args[0]);
    if (errors.length) {
      for (const e of errors) process.stderr.write(`FAIL: ${e}\n`);
      process.exit(1);
    }
    process.stdout.write(`OK: roles.json integrity verified against ${args[0]}\n`);
  },
};

// Commands that don't read roles.json — they work on a settings file alone, so they
// must run even when resolve.mjs is installed away from roles.json (e.g. under hooks/).
const NO_ROLES_DATA = new Set(["prune"]);

function main() {
  const [cmd, ...args] = process.argv.slice(2);
  if (!cmd || !(cmd in CLI)) {
    die(`usage: resolve.mjs <${Object.keys(CLI).join("|")}> [args]`, 2);
  }
  const data = NO_ROLES_DATA.has(cmd) ? {} : loadRoles();
  CLI[cmd](data, args);
}

// Run as CLI only when invoked directly (not when imported). Compare case-insensitively
// on Windows, where the invoking path's drive-letter / casing may differ from the
// module URL's even when they point at the same file.
function samePath(a, b) {
  const x = resolvePath(a);
  const y = resolvePath(b);
  return process.platform === "win32" ? x.toLowerCase() === y.toLowerCase() : x === y;
}
if (process.argv[1] && samePath(process.argv[1], fileURLToPath(import.meta.url))) {
  main();
}
